lib/page-loader.js

```diff
@@ -0,0 +1,104 @@
+/* global window, document */
+import mitt from 'mitt'
+
+const webpackModule = module
+
+export default class PageLoader {
+  constructor (buildId) {
+    this.buildId = buildId
+    this.pageCache = {}
+    this.pageLoadedHandlers = {}
+    this.registerEvents = mitt()
+    this.loadingRoutes = {}
+  }
+
+  normalizeRoute (route) {
+    if (route[0] !== '/') {
+      throw new Error('Route name should start with a "/"')
+    }
+
+    return route.replace(/index$/, '')
+  }
+
+  loadPage (route) {
+    route = this.normalizeRoute(route)
+
+    const cachedPage = this.pageCache[route]
+    if (cachedPage) {
+      return new Promise((resolve, reject) => {
+        if (cachedPage.error) return reject(cachedPage.error)
+        return resolve(cachedPage.page)
+      })
+    }
+
+    return new Promise((resolve, reject) => {
+      const fire = ({ error, page }) => {
+        this.registerEvents.off(route, fire)
+
+        if (error) {
+          reject(error)
+        } else {
+          resolve(page)
+        }
+      }
+
+      this.registerEvents.on(route, fire)
+
+      // Load the script if not asked to load yet.
+      if (!this.loadingRoutes[route]) {
+        this.loadScript(route)
+        this.loadingRoutes[route] = true
+      }
+    })
+  }
+
+  loadScript (route) {
+    route = this.normalizeRoute(route)
+
+    const script = document.createElement('script')
+    const url = `/_next/${encodeURIComponent(this.buildId)}/page${route}`
+    script.src = url
+    script.type = 'text/javascript'
+    script.onerror = () => {
+      const error = new Error(`Error when loading route: ${route}`)
+      this.registerEvents.emit(route, { error })
+    }
```

Comment:
I'd like to keep the next pages inside a variable of PageLoader. Anyway doing following is not exactly possible.

```
const newContent = `
  window.__NEXT_PAGES_[${JSON.stringify(routename)}] = {
    statusCode: ${statusCode},
    page: ${content}
  }
`
```
That's because, webpack is not available when the page is loading at first.
But I did the next best thing.